lib/cache/OperationWriter.js

```diff
@@ -0,0 +1,255 @@
+const {config, util: {BedrockError}} = require('bedrock');
+const {LruCache} = require('@digitalbazaar/lru-memoize');
+
+const PROMISE_QUEUE = new LruCache({max: 10, disposeOnSettle: true});
+
+// Error Messages
+const DATABASE_QUEUE_LENGTH_ERROR =
+  'Non-critical error: failed to cache length of operation queue on disk.';
+
+class OperationWriter {
+  constructor({ledgerNodeId, highWaterMark = 100, timeout = 1000} = {}) {
+    if(!ledgerNodeId) {
+      throw new Error(`"ledgerNodeId" is required.`);
+    }
+    this.ledgerNodeId = ledgerNodeId;
+    // maximum number of operations to have in the queue before there is
+    // pressure to schedule flushing the queue
+    this.highWaterMark = highWaterMark;
+    // the maximum amount of time that an operation can sit in the queue
+    // before being flushed when the operation writer is running
+    this.timeout = timeout;
+    // whether or not the operation writer is running
+    this.running = false;
+    // an ID for when the next timeout-based flush will occur
+    this.timeoutId = null;
+    // the queue of operations
+    this.queue = [];
+    // current stopped promise; resolves once stopped
+    this.stopped = null;
+    this._resolveStopped = null;
+    // tracks operation hashes for duplicate prevention
+    this.recentOperationHashes = new LRU({max: 1000, maxAge: 5 * 60 * 1000});
+    // operation queue collection
+    this.collection = database.collections.continuity2017_operation_queue;
+    // queue length in database
+    this.databaseOperationQueueLength = 0;
+    const operationsConfig = config['ledger-consensus-continuity'].operations;
+    // max operation queue length
+    this.MAX_OPERATION_QUEUE_LENGTH = operationsConfig.maxQueueSize;
+  }
+
+  /**
+   * Starts this OperationWriter. Any operations that have been added to its
+   * queue will be bulk inserted when the queue fills or when a timeout occurs.
+   */
+  start() {
+    const {queue, running, highWaterMark, timeout} = this;
+    if(running) {
+      // already running
+      return;
+    }
+
+    // now running
+    this.running = true;
+
+    // if the queue is at or over the highWaterMark, schedule immediate flush
+    if(queue.length >= highWaterMark) {
+      this._scheduleFlush({timeout: 0});
+    } else if(queue.length > 0) {
+      // queue is not empty, schedule a timeout flush
+      this._scheduleFlush({timeout});
+    }
+
+    this.stopped = new Promise(res => {
+      this._resolveStopped = res;
+    });
+  }
+
+  /**
+   * Stops this OperationWriter. Events can still be added to the queue, but
+   * they will not be automatically bulk inserted.
+   */
+  stop() {
+    const {running, timeoutId} = this;
+    if(!running) {
+      // already not running
+      return;
+    }
+    // cancel any scheduled flush
+    if(timeoutId !== null) {
+      clearTimeout(timeoutId);
+      this.timeoutId = null;
+    }
+    this.running = false;
+    this.stopped = null;
+    this._resolveStopped();
```

Comment:
I'm just skimming over most of the stop mechanics here and presuming it came from another place that is reasonably well tested or what we have here has some decent tests or something. It's probably not so critical that if it fails it will create irreparable harm to a peer's membership in a ledger.